kuma/wiki/content.py

```diff
@@ -1089,11 +1090,26 @@ def __iter__(self):
                 attrs = dict(token['data'])
                 for (namespace, name), value in attrs.items():
                     if name == 'src' and value:
-                        if not re.search(self.hosts, value):
+                        if not self.validate_src(value):
                             attrs[(namespace, 'src')] = ''
                     token['data'] = attrs
                 yield token
             if token['type'] == 'EndTag' and token['name'] == 'iframe':
                 in_iframe = False
             if not in_iframe:
                 yield token
+
+    def validate_src(self, src):
+        """Validate an iframe src against the allowed patterns."""
+        parts = urlsplit(src)
+        for scheme, netloc, path in self.allowed_src_patterns:
+            if parts.netloc != netloc or parts.scheme != scheme:
+                continue
+            if isinstance(path, re._pattern_type):
```

Comment:
As an alternative to using a non-public attribute of `re`, what do you think of something like this instead?
```python
if hasattr(path, 'match'):
```
or maybe something more strict, like this:
```python
if callable(getattr(path, 'match', None)):
```
or maybe this is more readable:
```python
if hasattr(path, 'match') and callable(path.match):
```

If you don't like any of those, that's fine. We can keep it like this.

Reply:
``if hasattr(path, 'match'):`` looks good to me. I've also seen this, which seems ugly:

```
if isinstance(path, type(re.compile(''))):
```